runtime/src/in_mem_accounts_index.rs

```diff
@@ -1258,6 +1259,69 @@ impl Drop for FlushGuard<'_> {
     }
 }
 
+/// Disable (and safely enable) the background flusher from evicting entries from the in-mem
+/// accounts index.  When disabled, no entries may be evicted.  When enabled, only eligible entries
+/// may be evicted (i.e. those not in a held range).
+///
+/// An RAII implementation of a scoped lock for the `stop_evictions` atomic flag/counter in
+/// `InMemAccountsIndex`.  When this structure is dropped (falls out of scope), the counter will
+/// decrement and conditionally notify its storage.
+///
+/// After successfully locking (calling `EvictionsGuard::lock()`), pass a reference to the
+/// `EvictionsGuard` instance to any function/code that requires `stop_evictions` to be
+/// incremented/decremented correctly.
+#[derive(Debug)]
+struct EvictionsGuard<'a> {
+    /// The number of active callers disabling evictions
+    stop_evictions: &'a AtomicU64,
+    /// The number of times that evictions have been disabled or enabled
+    num_state_changes: &'a AtomicU64,
+    /// Who will be notified after the evictions are re-enabled
+    storage_notifier: &'a WaitableCondvar,
+}
+
+impl<'a> EvictionsGuard<'a> {
+    #[must_use = "if unused, this evictions lock will be immediately unlocked"]
+    fn lock<T: IndexValue>(in_mem_accounts_index: &'a InMemAccountsIndex<T>) -> Self {
+        Self::lock_with(
+            &in_mem_accounts_index.stop_evictions,
+            &in_mem_accounts_index.stop_evictions_changes,
+            &in_mem_accounts_index.storage.wait_dirty_or_aged,
+        )
+    }
+
+    #[must_use = "if unused, this evictions lock will be immediately unlocked"]
+    fn lock_with(
+        stop_evictions: &'a AtomicU64,
+        num_state_changes: &'a AtomicU64,
+        storage_notifier: &'a WaitableCondvar,
+    ) -> Self {
+        num_state_changes.fetch_add(1, Ordering::Release);
+        stop_evictions.fetch_add(1, Ordering::Release);
+
+        Self {
+            stop_evictions,
+            num_state_changes,
+            storage_notifier,
+        }
+    }
+}
+
+impl Drop for EvictionsGuard<'_> {
+    fn drop(&mut self) {
+        let previous_value = self.stop_evictions.fetch_sub(1, Ordering::AcqRel);
+        debug_assert!(previous_value > 0);
+
+        let should_notify = previous_value == 1;
+        if should_notify {
+            // stop_evictions went to 0, so this bucket could now be ready to be aged
+            self.storage_notifier.notify_one();
+        }
+
+        self.num_state_changes.fetch_add(1, Ordering::Release);
```

Comment:
should this be  Ordering::AcqRel?

Reply:
We aren't keeping the returned/loaded value, so I don't think we need AcqRel here.

Reply:
If we also need to read/use the loaded value, then with Relaxed we wouldn't be guaranteed when the _visibility_ of both stores would happen (but there would still indeed be two increments).

Reply:
Then, Can we used `Ordering::Relaxed` here since there is no load or store for self.num_state_changes afterward?

Reply:
`num_state_changes` is loaded via `get_stop_evictions_changes()` (with `Acquire`) and is used in `evict_from_cache()` to test if things have finished. So we need the Happens-Before relationship that is provided with Acquire + Release (but is not provided with Relaxed).

Sort of related to https://github.com/solana-labs/solana/pull/25847#discussion_r892783847